File: server/src/routes/voice.ts
import { FastifyInstance } from "fastify";
import { SarvamSTTProvider } from "../ai/voice/sarvam-stt.js";
import { SarvamTTSProvider } from "../ai/voice/sarvam-tts.js";
import { config } from "../config/index.js";

export async function voiceRoutes(app: FastifyInstance) {
  const stt = new SarvamSTTProvider();
  const tts = new SarvamTTSProvider();

  app.get("/status", async () => ({
    stt: { provider: "sarvam", available: stt.isAvailable(), mockMode: !stt.isAvailable() },
    tts: { provider: "sarvam", available: tts.isAvailable(), mockMode: !tts.isAvailable() },
    languages: ["en", "hi", "mr"],
  }));

  app.post("/stt", async (request, reply) => {
    const parts = await request.file();
    if (!parts) {
      return reply.status(400).send({ error: "No audio file provided" });
    }

    const chunks: Buffer[] = [];
    for await (const chunk of parts.file) {
      chunks.push(chunk);
    }
    const audioBuffer = Buffer.concat(chunks);
    const language = (parts.fields as any)?.language?.value || "en";

    if (!stt.isAvailable()) {
      return {
        mockMode: true,
        text: "[Mock STT - Sarvam API key not configured]",
        confidence: 0.9,
        language,
      };
    }

    try {
      const result = await stt.transcribe(audioBuffer, language);
      return result;
    } catch (err: any) {
      return reply.status(500).send({ error: err.message });
    }
  });

  app.post("/tts", async (request, reply) => {
    const { text, language } = request.body as { text: string; language?: string };

    if (!text) {
      return reply.status(400).send({ error: "Text is required" });
    }

    if (!tts.isAvailable()) {
      return {
        mockMode: true,
        message: "Sarvam TTS mock mode",
        audio: null,
        format: "audio/wav",
      };
    }

    try {
      const result = await tts.synthesize(text, language || "en");
      reply.header("Content-Type", result.format);
      reply.header("X-Audio-Duration-Ms", result.durationMs.toString());
      return result.audioBuffer;
    } catch (err: any) {
      return reply.status(500).send({ error: err.message });
    }
  });
}

File: server-python/routes/voice.py
"""Voice routes — STT and TTS endpoints."""
from __future__ import annotations
import base64
from fastapi import APIRouter, HTTPException, Request
from providers import get_stt_provider, get_tts_provider
import config

router = APIRouter(prefix="/api/voice", tags=["voice"])


@router.get("/status")
async def voice_status():
    has_sarvam = bool(config.SARVAM_API_KEY)
    return {
        "stt": {"provider": "sarvam", "available": has_sarvam, "mockMode": not has_sarvam},
        "tts": {"provider": "sarvam", "available": has_sarvam, "mockMode": not has_sarvam},
        "languages": ["en", "hi", "mr"],
    }


@router.get("/voices")
async def list_voices():
    from providers.tts.sarvam_provider import SARVAM_VOICES
    return {"voices": SARVAM_VOICES}


@router.post("/stt")
async def speech_to_text(request: Request):
    try:
        body = await request.json()
        audio_b64 = body.get("audio", "")
        language = body.get("language", "en")

        if not audio_b64:
            raise HTTPException(status_code=400, detail="No audio provided")

        audio_data = base64.b64decode(audio_b64)
        stt = get_stt_provider()
        result = await stt.transcribe(audio_data, language)
        return result
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Speech-to-text failed")


@router.post("/tts")
async def text_to_speech(request: Request):
    try:
        body = await request.json()
        text = body.get("text", "")
        language = body.get("language", "en")
        speaker = body.get("speaker", "shubh")

        if not text:
            raise HTTPException(status_code=400, detail="No text provided")

        tts = get_tts_provider()
        audio = await tts.synthesize(text, language, speaker=speaker)
        if audio:
            return {
                "audio": base64.b64encode(audio).decode(),
                "format": "audio/wav",
                "mockMode": False,
            }
        raise HTTPException(status_code=502, detail="TTS returned empty audio")
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Text-to-speech failed")
